### PR TITLE
ENH: allow single input argument in numpy.broadcast

### LINKED ISSUE #6899: Why does `numpy.broadcast` not accept a single input array?
I'm wondering what's the reasoning behind forbidding a single array as input to `numpy.broadcast`. When writing code which is supposed to work in N dimensions with N >= 1, it would be nice to be able to use `broadcast` as-is. A standard use case is the calculation of the shape of an output array when using meshgrids as input for function evaluation:
```python
>>> coord_vecs = [[1, 2], [3, 4, 5]]
>>> mesh = np.meshgrid(*coord_vecs, indexing='ij', sparse=True)  # for large stuff
>>> out_shape = np.broadcast(*mesh).shape
>>> out_shape
(2, 3)
>>> coord_vecs = [[1, 2]]
>>> mesh = np.meshgrid(*coord_vecs, indexing='ij', sparse=True)  # still works
>>> out_shape = np.broadcast(*mesh).shape
Traceback (most recent call last):
...
ValueError: Need at least two and fewer than (32) array objects.
```
As noted above, `meshgrid` works and returns a list with one element, as expected. Why does `broadcast` not return a broadcast object with shape `(2,)` in the example above? The other public methods of such an object would probably also make sense with a single input array.

EDIT: Last sentence was wrong in the first version.

### PR DESCRIPTION
**Pull request by kohr-h**
Fixes #6899. The [documentation](https://github.com/numpy/numpy/blob/d2435303f298eaae9118b17588d1db8c7144bb0f/numpy/add_newdocs.py#L467-L482) does not mention any restriction, so I guess there is no need to adapt it?

### CODE REVIEW

**Comment by jaimefrio** [comment]
This change requires some tests to validate that it works.

Also, there are three functions:
 * `PyArray_MultiIterFromObjects`
 * `PyArray_MultiIterNew`
 * `arraymultiter_new`

doing almost the exact same thing, and this change brings them slightly apart, which is not a good thing.

Probably the right thing to do is to refactor the code so that there is a single implementation of the multi-iterator constructor that gets called by the others.  But at the very least, the change should be made on all three.

**Comment by kohr-h** [comment]
> This change requires some tests to validate that it works.

Where would such a test go? Would it be sufficient to just always use `np.broadcast` in [this test](https://github.com/numpy/numpy/blob/5dab9275475e70630d7b1242c3530a6726c61fa2/numpy/core/tests/test_indexing.py#L897-L901), assuming there is also a single-array test case?



> Also, there are three functions:
> 
>     PyArray_MultiIterFromObjects
>     PyArray_MultiIterNew
>     arraymultiter_new
> 
> doing almost the exact same thing, and this change brings them slightly apart, which is not a good thing.
> 
> Probably the right thing to do is to refactor the code so that there is a single implementation of the multi-iterator constructor that gets called by the others. But at the very least, the change should be made on all three.

I can easily do the least and adapt those three functions. Tests are also relevant here, I guess. 
Regarding the common code, I assume you are referring (roughly) to [this portion](https://github.com/numpy/numpy/blob/5dab9275475e70630d7b1242c3530a6726c61fa2/numpy/core/src/multiarray/iterators.c#L1456-L1475) of the constructor. Do you want me to make an attempt? How about names for small helpers then, any suggestion?

**Comment by kohr-h** [comment]
I just found the dedicated `TestBroadcast` case.

**Comment by kohr-h** [comment]
Single-arg broadcast has a test now. I did not find a nice way to modify the existing list and get the desired result, so I added a few extra lines.

**Comment by kohr-h** [comment]
I took the easy route and just changed numbers and error messages three times. The "proper" code refactoring goes beyond my knowledge of the C part.
`broadcast` does not seem to be used very much in the rest of the library, I just found one place in `stride_tricks.py` where a distinction had to be made between `len(args) == 1` and all other cases. This check has been removed.

Question: are there dedicated tests for the `PyArray_MultiIterFromObjects, PyArray_MultiIterNew,
arraymultiter_new` functions where a case for length 1 could be added?

File: numpy/core/tests/test_numeric.py [hunk 1]
```diff
@@ -2207,11 +2207,19 @@ def test_broadcast_in_args(self):
             for a, ia in zip(arrs, mit.iters):
                 assert_(a is ia.base)
 
+        # gh-6899
```
Comment (jaimefrio) [inline numpy/core/tests/test_numeric.py]: I think this deserves its own test method, `test_broacast_single_arg` or something along that line.

**Comment by kohr-h** [comment]
Rebased and streamlined the commits. Any further action necessary?

**Comment by jaimefrio** [comment]
One last request: I think this is worthy of a mention in the release notes.

**Comment by kohr-h** [comment]
Sure, under which heading? Changes?

**Comment by kohr-h** [comment]
Looking ok?

**Comment by jaimefrio** [comment]
Thanks Holger, in it goes.

**Comment by kohr-h** [comment]
My pleasure!

**Pull request merged**